README.md
![SS logo](https://raw.githubusercontent.com/maliceio/malice-shadow-server/master/logo.png)

malice-shadow-server
====================

[![Circle CI](https://circleci.com/gh/maliceio/malice-shadow-server.png?style=shield)](https://circleci.com/gh/maliceio/malice-shadow-server) [![License](http://img.shields.io/:license-mit-blue.svg)](http://doge.mit-license.org) [![Docker Stars](https://img.shields.io/docker/stars/malice/shadow-server.svg)](https://hub.docker.com/r/malice/shadow-server/) [![Docker Pulls](https://img.shields.io/docker/pulls/malice/shadow-server.svg)](https://hub.docker.com/r/malice/shadow-server/) [![Docker Image](https://img.shields.io/badge/docker image-19.43 MB-blue.svg)](https://hub.docker.com/r/malice/shadow-server/)

Malice ShadowServer Hash Lookup Plugin

This repository contains a **Dockerfile** of the [ShadowServer](https://www.shadowserver.org/wiki/) malice plugin **malice/shadow-server**.

### Dependencies

-	[malice/alpine:tini](https://hub.docker.com/r/malice/alpine/)

### Installation

1.	Install [Docker](https://www.docker.io/).
2.	Download [trusted build](https://hub.docker.com/r/malice/shadow-server/) from public [DockerHub](https://hub.docker.com): `docker pull malice/shadow-server`

### Usage

```
docker run --rm malice/shadow-server MD5/SHA1
```

```bash
Usage: shadow-server [OPTIONS] COMMAND [arg...]

Malice ShadowServer Hash Lookup Plugin

Version: v0.1.0, BuildTime: 20160219

Author:
  blacktop - <https://github.com/blacktop>

Options:
  --verbose, -V      verbose output
  --elasitcsearch value  elasitcsearch address for Malice to store results [$MALICE_ELASTICSEARCH]
  --post, -p         POST results to Malice webhook [$MALICE_ENDPOINT]
  --proxy, -x        proxy settings for Malice webhook endpoint [$MALICE_PROXY]
  --table, -t        output as Markdown table
  --help, -h         show help
  --version, -v      print the version

Commands:
  help	Shows a list of commands or help for one command

Run 'shadow-server COMMAND --help' for more information on a command.
```

This will output to stdout and POST to malice results API webhook endpoint.

### Sample Output **sandbox** JSON:

```json
{
  "shadow-server": {
    "found": true,
    "sandbox": {
      "md5": "aca4aad254280d25e74c82d440b76f79",
      "sha1": "6fe80e56ad4de610304bab1675ce84d16ab6988e",
      "first_seen": "2010-06-15 03:09:41",
      "last_seen": "2010-06-15 03:09:41",
      "type": "exe",
      "ssdeep": "12288:gOqOB0v2eZJys73dOvXDpNjNe8NuMpX4aBaa48L/93zKnP6ppgg2HFZlxVPbZX:sOA2eZJ8NI8Nah8L/4PqmTVPlX",
      "antivirus": {
        "AVG7": "Downloader.Generic9.URM",
        "AntiVir": "WORM/VB.NVA",
        "Avast-Commercial": "Win32:Zbot-LRA",
        "Clam": "Trojan.Downloader-50691",
        "DrWeb": "Win32.HLLW.Autoruner.6014",
        "F-Prot6": "W32/Worm.BAOX",
        "F-Secure": "Worm:W32/Revois.gen!A",
        "G-Data": "Trojan.Generic.2609117",
        "Ikarus": "Trojan-Downloader.Win32.VB",
        "Kaspersky": "Trojan.Win32.Cosmu.nyl",
        "McAfee": "Generic",
        "NOD32": "Win32/AutoRun.VB.JP",
        "Norman": "Suspicious_Gen2.SKLJ",
        "Panda": "W32/OverDoom.A",
        "QuickHeal": "Worm.VB.at",
        "Sophos": "Troj/DwnLdr-HQY",
        "TrendMicro": "TROJ_DLOADR.SMM",
        "VBA32": "Trojan.VBO.011858",
        "Vexira": "Trojan.DL.VB.EEDT",
        "VirusBuster": "Worm.VB.FMYJ"
      }
    },
    "whitelist": null
  }
}
```

### Sample Output **whitelist** JSON:

```json
{
  "shadow-server": {
    "found": true,
    "sandbox": {
      "md5": "5e28284f9b5f9097640d58a73d38ad4c",
      "sha1": "7a90f8b051bc82cc9cadbcc9ba345ced02891a6c",
      "first_seen": "2009-07-24 02:09:53",
      "last_seen": "2009-07-24 02:09:53",
      "type": "exe",
      "ssdeep": "1536:bwOnbNQKLjWDyy1o5I0foMJUEbooPRrKKReFX3:RNQKPWDyDI0fFJltZrpReFX3",
      "antivirus": {}
    },
    "whitelist": {
      "application_type": "exe",
      "binary": "1",
      "bit": "32",
      "crc32": "877EA041",
      "description": "Notepad",
      "dirname": "c:\\WINDOWS\\system32",
      "filename": "notepad.exe",
      "filesize": "69120",
      "filetimestamp": "04/14/2008 12:00:00",
      "fileversion": "5.1.2600.5512",
      "language": "English",
      "language_code": "1033",
      "md5": "5E28284F9B5F9097640D58A73D38AD4C",
      "media_source": "http://www.microsoft.com/",
      "mfg_name": "Microsoft Corporation",
      "os_mfg": "Microsoft Corporation",
      "os_name": "Microsoft Windows XP Professional Service Pack 3 (build 2600)",
      "os_version": "5.1",
      "product_name": "Microsoft Windows Operating System",
      "product_version": "5.1.2600.5512",
      "reference": "os_patches_all",
      "sha1": "7A90F8B051BC82CC9CADBCC9BA345CED02891A6C",
      "sha256": "865F34FE7BA81E9622DDBDFC511547D190367BBF3DAD21CEB6DA3EEC621044F5",
      "sha512": "CB7218CFEA8813AE8C7ACF6F7511AECBEB9D697986E0EB8538065BF9E3E9C6CED9C29270EB677F5ACF08D2E94B21018D8C4A376AA646FA73CE831FC87D448934",
      "sig_timestamp": "04/14/2008 02:07:47",
      "sig_trustfile": "C:\\WINDOWS\\system32\\CatRoot\\{F750E6C3-38EE-11D1-85E5-00C04FC295EE}\\NT5.CAT",
      "signer": "Microsoft Windows Component Publisher",
      "source": "AppInfo",
      "source_version": "1.8",
      "strongname_signed": "0",
      "trusted_signature": "1"
    }
  }
}
```

### Sample Output **whitelist** (Markdown Table):

---

#### shadow-server

##### WhiteList

| Found | Filename    | Description | ProductName                        |
|-------|-------------|-------------|------------------------------------|
| true  | notepad.exe | Notepad     | Microsoft Windows Operating System |

---

### Sample Output **sandbox** (Markdown Table):

---

#### shadow-server

##### AntiVirus

-	FirstSeen: 6/15/2010 3:09AM
-	LastSeen: 6/15/2010 3:09AM

| Vendor           | Signature                  |
|------------------|----------------------------|
| F-Prot6          | W32/Worm.BAOX              |
| G-Data           | Trojan.Generic.2609117     |
| NOD32            | Win32/AutoRun.VB.JP        |
| Avast-Commercial | Win32:Zbot-LRA             |
| DrWeb            | Win32.HLLW.Autoruner.6014  |
| Norman           | Suspicious_Gen2.SKLJ       |
| Panda            | W32/OverDoom.A             |
| Vexira           | Trojan.DL.VB.EEDT          |
| VirusBuster      | Worm.VB.FMYJ               |
| AntiVir          | WORM/VB.NVA                |
| Clam             | Trojan.Downloader-50691    |
| Ikarus           | Trojan-Downloader.Win32.VB |
| Kaspersky        | Trojan.Win32.Cosmu.nyl     |
| QuickHeal        | Worm.VB.at                 |
| VBA32            | Trojan.VBO.011858          |
| AVG7             | Downloader.Generic9.URM    |
| McAfee           | Generic                    |
| Sophos           | Troj/DwnLdr-HQY            |
| TrendMicro       | TROJ_DLOADR.SMM            |
| F-Secure         | Worm:W32/Revois.gen!A      |

---

Documentation
-------------

### To write results to [ElasticSearch](https://www.elastic.co/products/elasticsearch)

```bash
$ docker volume create --name malice
$ docker run -d --name elastic \
                -p 9200:9200 \
                -v malice:/usr/share/elasticsearch/data \
                 blacktop/elasticsearch
$ docker run --rm --link elastic malice/shadow-server -t MD5/SHA1
```

### Issues

Find a bug? Want more features? Find something missing in the documentation? Let me know! Please don't hesitate to [file an issue](https://github.com/maliceio/malice-shadow-server/issues/new) and I'll get right on it.

### CHANGELOG

See [`CHANGELOG.md`](https://github.com/maliceio/malice-shadow-server/blob/master/CHANGELOG.md)

### Contributing

[See all contributors on GitHub](https://github.com/maliceio/malice-shadow-server/graphs/contributors).

Please update the [CHANGELOG.md](https://github.com/maliceio/malice-shadow-server/blob/master/CHANGELOG.md) and submit a [Pull Request on GitHub](https://help.github.com/articles/using-pull-requests/).

### License

MIT Copyright (c) 2016 **blacktop**


CHANGELOG.md
Change Log
==========

All notable changes to this project will be documented in this file.

[Unreleased][unreleased]
------------------------

### Fixed

### Added

### Removed

### Changed

[v0.1.0] - 2016-09-05
---------------------

### Fixed

-	ShadowServer changed the output for Not found items.

### Added

### Removed

### Changed

-	Upgrade golang to 1.7.3


shadow-server.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"strings"
	"time"

	log "github.com/Sirupsen/logrus"
	"github.com/crackcomm/go-clitable"
	"github.com/fatih/structs"
	"github.com/levigross/grequests"
	"github.com/maliceio/go-plugin-utils/database/elasticsearch"
	"github.com/maliceio/go-plugin-utils/utils"
	"github.com/parnurzeal/gorequest"
	"github.com/urfave/cli"
)

// Version stores the plugin's version
var Version string

// BuildTime stores the plugin's build time
var BuildTime string

const (
	name     = "shadow-server"
	category = "intel"
)

// ShadowServer json object
type ShadowServer struct {
	Results ResultsData `json:"shadow-server"`
}

// ResultsData json object
type ResultsData struct {
	Found     bool             `json:"found" structs:"found"`
	SandBox   SandBoxResults   `json:"sandbox" structs:"sandbox"`
	WhiteList WhiteListResults `json:"whitelist" structs:"whitelist"`
}

// SandBoxResults is a shadow-server SandboxApi results JSON object
type SandBoxResults struct {
	MetaData  map[string]string `json:"metadata,omitempty" structs:"metadata,omitempty"`
	Antivirus map[string]string `json:"antivirus" structs:"antivirus"`
}

type sandBoxMetaData struct {
	MD5       string    `json:"md5" structs:"md5"`
	SHA1      string    `json:"sha1" structs:"sha1"`
	FirstSeen time.Time `json:"first_seen" structs:"first_seen"`
	LastSeen  time.Time `json:"last_seen" structs:"last_seen"`
	FileType  string    `json:"type" structs:"type"`
	SSDeep    string    `json:"ssdeep" structs:"ssdeep"`
}

// WhiteListResults is a shadow-server bin-test results JSON object
type WhiteListResults map[string]string

// IsEmpty checks if ResultsData is empty
func (r ResultsData) IsEmpty() bool {
	return reflect.DeepEqual(r, ResultsData{})
}

func hashType(hash string) *grequests.RequestOptions {
	hashTyp, err := utils.GetHashType(hash)
	if err != nil {
		return &grequests.RequestOptions{}
	}

	return &grequests.RequestOptions{Params: map[string]string{hashTyp: hash}}
}

func parseWhiteListOutput(whitelistout string) WhiteListResults {
	whitelist := WhiteListResults{}

	lines := strings.Split(whitelistout, "\n")

	if len(lines) > 1 {

		fields := strings.SplitN(lines[0], " ", 2)

		if len(fields) == 2 {
			if fields[1] == "" {
				return nil
			}
			utils.Assert(json.Unmarshal([]byte(fields[1]), &whitelist))
		}
	}

	// fmt.Printf("%#v\n", whitelist)
	return whitelist
}

// whiteListHash test hash against a list of known software applications
func whiteListHash(hash string) WhiteListResults {

	resp, err := grequests.Get("http://bin-test.shadowserver.org/api", hashType(hash))

	if err != nil {
		log.Fatalln("Unable to make request: ", err)
	}

	if resp.Ok != true {
		log.Println("Request did not return OK")
	}

	return parseWhiteListOutput(resp.String())
}

func parseSandboxAPIOutput(sandboxapiout string) SandBoxResults {
	var sandbox SandBoxResults

	lines := strings.Split(sandboxapiout, "\n")

	if len(lines) == 1 {
		if strings.Contains(lines[0], "! No match found") {
			return sandbox
		}
		if strings.Contains(lines[0], "! Whitelisted:") {
			return sandbox
		}
	}

	if len(lines) == 2 {
		values := strings.Split(lines[0], ",")
		if len(values) == 6 {
			// "2009-07-24 02:09:53"
			const longForm = "2006-01-02 15:04:05"
			timeFirstSeen, _ := time.Parse(longForm, strings.Trim(values[2], "\""))
			timeLastSeen, _ := time.Parse(longForm, strings.Trim(values[3], "\""))
			meta := make(map[string]string)
			meta["md5"] = strings.Trim(values[0], "\"")
			meta["sha1"] = strings.Trim(values[1], "\"")
			meta["first_seen"] = timeFirstSeen.String()
			meta["last_seen"] = timeLastSeen.String()
			meta["type"] = strings.Trim(values[4], "\"")
			meta["ssdeep"] = strings.Trim(values[5], "\"")
			sandbox = SandBoxResults{MetaData: meta}
		}
		if len(lines[1]) == 2 {
			sandbox.Antivirus = nil
		} else {
			utils.Assert(json.Unmarshal([]byte(lines[1]), &sandbox.Antivirus))
		}
	}

	return sandbox
}

// sandboxAPISearch search hash in AV results
func sandboxAPISearch(hash string) SandBoxResults {

	// NOTE: https://godoc.org/github.com/levigross/grequests
	ro := &grequests.RequestOptions{
		Params: map[string]string{
			"query": hash,
		},
	}
	resp, err := grequests.Get("http://innocuous.shadowserver.org/api/", ro)

	if err != nil {
		log.Fatalln("Unable to make request: ", err)
	}

	if resp.Ok != true {
		log.Println("Request did not return OK")
	}

	// fmt.Println(resp.String())
	return parseSandboxAPIOutput(resp.String())
}

// LookupHash retreieves the shadow-server file report for the given hash
func LookupHash(hash string) ResultsData {

	lookup := ResultsData{}

	lookup.WhiteList = whiteListHash(hash)
	lookup.SandBox = sandboxAPISearch(hash)

	if lookup.IsEmpty() {
		lookup.Found = false
	} else {
		lookup.Found = true
	}
	// fmt.Printf("%#v", lookup)
	return lookup
}

func printTableFormattedTime(t string) string {
	timeInTableFormat, _ := time.Parse("2006-01-02 15:04:05 -0700 UTC", t)
	return timeInTableFormat.Format("1/02/2006 3:04PM")
}

func printMarkDownTable(ss ShadowServer) {
	fmt.Println("#### ShadowServer")
	if ss.Results.WhiteList != nil {
		fmt.Println("##### WhiteList")
		table := clitable.New([]string{"Found", "Filename", "Description", "ProductName"})
		table.AddRow(map[string]interface{}{
			"Found":       ss.Results.Found,
			"Filename":    ss.Results.WhiteList["filename"],
			"Description": ss.Results.WhiteList["description"],
			"ProductName": ss.Results.WhiteList["product_name"],
		})
		table.Markdown = true
		table.Print()
	} else if ss.Results.SandBox.Antivirus != nil {
		fmt.Println("##### AntiVirus")
		// fmt.Printf(" - FirstSeen: %s\n", ss.Results.SandBox.MetaData["first_seen"].Format("1/02/2006 3:04PM"))
		fmt.Printf(" - FirstSeen: %s\n", printTableFormattedTime(ss.Results.SandBox.MetaData["first_seen"]))
		fmt.Printf(" - LastSeen: %s\n", printTableFormattedTime(ss.Results.SandBox.MetaData["last_seen"]))
		fmt.Println()
		table := clitable.New([]string{"Vendor", "Signature"})
		for key, value := range ss.Results.SandBox.Antivirus {
			table.AddRow(map[string]interface{}{"Vendor": key, "Signature": value})
		}
		table.Markdown = true
		table.Print()
	} else {
		fmt.Println(" - Not found")
	}
	fmt.Println()
}

func printStatus(resp gorequest.Response, body string, errs []error) {
	fmt.Println(body)
}

func main() {

	var elastic string

	cli.AppHelpTemplate = utils.AppHelpTemplate
	app := cli.NewApp()

	app.Name = "shadow-server"
	app.Author = "blacktop"
	app.Email = "https://github.com/blacktop"
	app.Version = Version + ", BuildTime: " + BuildTime
	app.Compiled, _ = time.Parse("20060102", BuildTime)
	app.Usage = "Malice ShadowServer Hash Lookup Plugin"
	app.Flags = []cli.Flag{
		cli.BoolFlag{
			Name:  "verbose, V",
			Usage: "verbose output",
		},
		cli.StringFlag{
			Name:        "elasitcsearch",
			Value:       "",
			Usage:       "elasitcsearch address for Malice to store results",
			EnvVar:      "MALICE_ELASTICSEARCH",
			Destination: &elastic,
		},
		cli.BoolFlag{
			Name:   "post, p",
			Usage:  "POST results to Malice webhook",
			EnvVar: "MALICE_ENDPOINT",
		},
		cli.BoolFlag{
			Name:   "proxy, x",
			Usage:  "proxy settings for Malice webhook endpoint",
			EnvVar: "MALICE_PROXY",
		},
		cli.BoolFlag{
			Name:  "table, t",
			Usage: "output as Markdown table",
		},
	}
	app.ArgsUsage = "MD5/SHA1 hash of file"
	app.Action = func(c *cli.Context) error {

		if c.Args().Present() {
			if c.Bool("verbose") {
				log.SetLevel(log.DebugLevel)
			}

			hash := c.Args().First()
			ssReport := LookupHash(hash)
			ss := ShadowServer{Results: ssReport}

			// upsert into Database
			elasticsearch.InitElasticSearch(elastic)
			elasticsearch.WritePluginResultsToDatabase(elasticsearch.PluginResults{
				ID:       utils.Getopt("MALICE_SCANID", hash),
				Name:     name,
				Category: category,
				Data:     structs.Map(ss.Results),
			})

			if c.Bool("table") {
				printMarkDownTable(ss)
			} else {
				ssJSON, err := json.Marshal(ss)
				utils.Assert(err)
				if c.Bool("post") {
					request := gorequest.New()
					if c.Bool("proxy") {
						request = gorequest.New().Proxy(os.Getenv("MALICE_PROXY"))
					}
					request.Post(os.Getenv("MALICE_ENDPOINT")).
						Set("X-Malice-ID", utils.Getopt("MALICE_SCANID", hash)).
						Send(string(ssJSON)).
						End(printStatus)

					return nil
				}
				fmt.Println(string(ssJSON))
			}
		} else {
			log.Fatal(fmt.Errorf("Please supply a MD5/SHA1 hash to query."))
		}
		return nil
	}

	err := app.Run(os.Args)
	utils.Assert(err)
}
